File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.cpp
#include<bits/stdc++.h>
using namespace std;
using ll = long long;

int solution(vector<vector<int>> board, vector<vector<int>> skill) {
    int n = board.size();
    int m = board[0].size();

    vector<vector<int>> diff(n+1, vector<int>(m+1, 0));
    
    for (auto s : skill){
        int t = s[0], r1= s[1], c1= s[2], r2= s[3], c2= s[4], d= s[5];
        
        if (t == 1) d = -d;

        diff[r1][c1] += d;
        diff[r1][c2+1] -= d;
        diff[r2+1][c1] -= d;
        diff[r2+1][c2+1] += d;
    }

    for (int i = 0; i < n; i++){
        for (int j = 1; j < m; j++) diff[i][j] += diff[i][j-1];
    }

    for (int j = 0; j < m; j++){
        for (int i = 1; i < n; i++) diff[i][j] += diff[i-1][j];
    }

    int answer = 0;
    for (int i = 0; i < n; i++){
        for (int j = 0; j < m; j++){
            if (board[i][j] + diff[i][j] > 0) answer++;
        }
    }
    return answer;
}

File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.py
def solution(board, skill):
    n = len(board)
    m = len(board[0])

    diff = [[0] * (m+1) for _ in range(n+1)]

    for s in skill:
        t, r1, c1, r2, c2, d = s
        
        if t == 1:
            d = -d

        diff[r1][c1] += d
        diff[r1][c2+1] -= d
        diff[r2+1][c1] -= d
        diff[r2+1][c2+1] += d
    
    for i in range(n):
        for j in range(1, m):
            diff[i][j] += diff[i][j-1]

    for j in range(m):
        for i in range(1, n):
            diff[i][j] += diff[i-1][j]
    answer = 0

    for i in range(n):
        for j in range(m):
            if board[i][j] + diff[i][j] > 0:
                answer += 1
            
    return answer